internal/netxlite/getaddrinfo_android.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2014 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build cgo && android
+
+package netxlite
+
+//#include <netdb.h>
+import "C"
+
+import (
+	"errors"
+	"syscall"
+)
+
+const getaddrinfoAIFlags = C.AI_CANONNAME
+
+// toError is the function that converts the return value from
+// the getaddrinfo function into a proper Go error.
+//
+// This function is adapted from cgoLookupIPCNAME
+// https://github.com/golang/go/blob/go1.17.6/src/net/cgo_unix.go#L145
+//
+// SPDX-License-Identifier: BSD-3-Clause.
+func (state *getaddrinfoState) toError(code C.int, err error) ([]string, error) {
+	switch code {
+	case C.EAI_SYSTEM:
+		if err == nil {
+			// err should not be nil, but sometimes getaddrinfo returns
+			// code == C.EAI_SYSTEM with err == nil on Linux.
+			// The report claims that it happens when we have too many
+			// open files, so use syscall.EMFILE (too many open files in system).
+			// Most system calls would return ENFILE (too many open files),
+			// so at the least EMFILE should be easy to recognize if this
+			// comes up again. golang.org/issue/6232.
+			err = syscall.EMFILE
+		}
+		return nil, newErrGetaddrinfo(int64(code), err)
+	case C.EAI_NONAME, C.EAI_NODATA:
+		// We have seen that on Android systems NXDOMAIN maps to
+		// EAI_NODATA and it's unclear whether this is the case for
+		// other systems as well. As far as we know, this does not
+		// happen for GNU/Linux, Windows, and macOS but it may be
+		// the case for some other systems (maybe BSD?).
+		//
+		// So the design choice here is to map to "no answer" for
+		// robustness but _also_ to record the original getaddrinfo
+		// code, so one can see it into the results.
+		//
+		// See https://github.com/ooni/probe/issues/2029 for the
+		// investigation on Android's getaddrinfo.
+		err = errors.New(DNSNoAnswerSuffix) // so it becomes FailureNSNoAnswer
```

Comment:
Good observation! I continued this work in another branch and what I ended up doing is creating a new error for this getaddrinfo retval. It turned out that Android returns `EAI_NODATA` for _any_ error. So, what we basically need to do is to flag Android's resolver as not reliable and only conclude there's NXDOMAIN when we see `EAI_NODATA` on Android and NXDOMAIN in the test helper. See https://github.com/ooni/probe/issues/2029#issuecomment-1140258729 for a complete discussion of how the Android resolver works and see https://github.com/ooni/probe-cli/pull/764 for the implementation.

Reply:
This comment is now going to disappear because I merged https://github.com/ooni/probe-cli/pull/764 into master and this file was not part of https://github.com/ooni/probe-cli/pull/764: I merged Android and Linux code together so I could run unit tests for Android code from Linux.

Reply:
Though, I suppose you'll probably see these messages in your inbox.

Reply:
Ah, good, the conversation is still there. I am just going to mark it as resolved. Reopen if you think it's proper.